Bluetooth: Audio: Serialize subscription after discovery

asbjornsabo opened the pull request:
The media control client used to subscribe to GMCS characteristics as
the characteristics were discovered, in effect queuing ATT requests.

Since then, the ATT MTU exchange has become automatic, leading to
larger ATT MTUs and more discovered characteristics per ATT response,
which again lead to the client enqueuing more subscriptions and
running out of buffers, with discovery failing as a result.

This PR moves the subscription to GMCS characteristics to a separate
sequence, so that it happens serially after discovery, rather than in 
parallel with the discovery.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/44579

asbjornsabo committed:
Bluetooth: Audio: Media - refactor include discovery start

Move the start of discovery of included services to a separate
function, in preparation for rewriting discovery.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: MCC - refactor subscription

Move subscription to characteristics to separate function.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: MCC - warn on subscription failure

Add warning if subscription to characteristics fails.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: MCC - serialized subscription after discovery

The media control client used to subscribe to GMCS characteristics as
the characteristics were discovered, in effect queuing ATT requests.

Since then, the ATT MTU exchange has become automatic, leading to
larger ATT MTUs and more discovered characteristics per ATT response,
which again lead to the client enqueuing more subscriptions and
running out of buffers, with discovery failing as a result.

This commit moves the subscription to GMCS characteristics to a
separate sequence, so that it happens serially after discovery, rather
than in parallel with the discovery.

- Remove separate discovery parameters for the characteristics - they
  are no longer needed.
- Add a subscription callback and a couple of subscription functions

The subscription for characteristics for the included OTS has not been
modified.  Only two of these are subscribed to, and doing that is
currently not a problem.  And I need to better understand the
interaction with OTS and how these subscriptipons are used.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Refactor subscription to get chaining into callback

Thalley commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1238,150 +1489,116 @@ static uint8_t discover_mcs_char_func(struct bt_conn *conn,
-		}
-
 		/* Continue to search for more attributes */
+		BT_DBG("RETURNING");
```
Remove? :)

asbjornsabo replied:
I thought I had ... :-)

Thalley commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1472,8 +1685,9 @@ int bt_mcc_init(struct bt_mcc_cb *cb)
  * 1: Discover GMCS primary service (started here)
  * 2: Discover characteristics of GMCS
+ * 3: Subscribe to characteristics of GMCS
  * 3: Discover OTS service included in GMCS
- * 4: Discover characteristics of OTS
+ * 4: Discover characteristics of OTS and subscribe to them
```
Can't help but to wonder that if we do these steps in this order, that the OTS characteristic discovery function should be split from the MCS discovery function, instead of guarding the individual characteristics with `#ifdef CONFIG_BT_MCC_OTS`. 

Not needed to be modified by this PR of course.

asbjornsabo replied:
Ah - this is the confusing part of MCS.
The OTS characteristic discovery function is split out as a separate function - `discover_otc_char_func()`
The individual characteristics that are guarded are MCS characteristics, but they are dependent upon there being an MCS.

Thalley replied:
Ah, right. So it's not as trivial to "fix".

asbjornsabo commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1049,6 +1032,15 @@ static uint8_t mcs_notify_handler(struct bt_conn *conn,
 	return BT_GATT_ITER_CONTINUE;
 }
 
+static void discovery_failure(struct bt_conn *conn, const char *errmsg, int err)
+{
+	BT_DBG("%s (err: %d)", errmsg, err);
```
But still keep the rest of the function, I assume?

It seemed natural to me to put the debug with the rest of the common handling.  It saves more than twenty debug statements elsewhere in the code. And the error message will tell where it was called from.
Not using the `errmsg` is a good point, though.
If you prefer the twenty debug statements, I will change it.

Thalley replied:
I do prefer the twenty debug statements actually (even though that sounds crazy ;) ), as that will provide more context. 

If DEBUG is turned off, are the static strings to the function calls still allocated?

asbjornsabo replied:
They will be, as far as I understand.  Another point in favour of keeping the BT_DEBUG local.
Will change.

asbjornsabo replied:
To clarify: The static strings that are there now will be allocated.  Any strings given to a BT_DEBUG macro I guess will not be.

asbjornsabo commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1195,19 +1182,283 @@ static uint8_t discover_include_func(struct bt_conn *conn,
+	}
+}
+
+static void subscribe_mcs_char_func(struct bt_conn *conn);
```
I think I have seen forward declarations placed close to the place of use/need several places in the stack, both for structs and functions.  (Looking now, I found one example in gatt.c).

To me it makes sense to have it with the other related functions - it is only there to sort out the dependencies between the three next functions, so it is all within forty lines of code, and easy to see.

Thalley replied:
No strong opinions nor any official guidelines on this :) So fine to keep as is

asbjornsabo commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1195,19 +1182,283 @@ static uint8_t discover_include_func(struct bt_conn *conn,
+static void subscription_cb(struct bt_conn *conn, uint8_t err,
+			    struct bt_gatt_subscribe_params *params)
+{
+	if (err) {
+		discovery_failure(conn, "Subscription callback error", err);
+	}
```
Fixed

asbjornsabo commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1195,19 +1182,283 @@ static uint8_t discover_include_func(struct bt_conn *conn,
+			 struct bt_gatt_subscribe_params *sub_params)
+{
+	/* With ccc_handle == 0 it will use auto discovery */
+	sub_params->ccc_handle = 0;
+	sub_params->end_handle = cur_mcs_inst->end_handle;
+	sub_params->value = BT_GATT_CCC_NOTIFY;
+	sub_params->value_handle = handle;
+	sub_params->notify = mcs_notify_handler;
+	sub_params->subscribe = subscription_cb;
+
```
That would seem natural, yes.  And it would save some lines of code.

The reason I like to have it where it currently is is that that pointer is also used as the flag for whether the characteristic has already been subscribed to or not.  Having the "setting" of the flag close to the testing of the flags helps understanding, IMO.
But a net saving of 16 lines of code trumps that, I guess.  Will change.

Thalley replied:
I get that, but you are also checking the handle which are set by this function, so the argument isn't really solid :)

asbjornsabo commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1195,19 +1182,283 @@ static uint8_t discover_include_func(struct bt_conn *conn,
+		if (err) {
+			discovery_failure(conn, "Could not subscribe", err);
+		}
+		return;
+	}
```
I did.
I think I chose to do it this way because of the #ifdef-s that break that pattern - they come in between the else and the following if.

Normally, we have
```
if (<some_condition>) {
        do_something()
} else if (<some_other_condition>) {
        do_something_else()
}
```
But here it would become
```
if (<some_condition>) {
        do_something()
} else 
#ifdef MCC_OTS
if (<some_other_condition>) {
        do_something_else()
}
```

Thalley replied:
Right, that makes sense. The numerous `#ifdef`'s would be a hinder for that. My suggestion from https://github.com/zephyrproject-rtos/zephyr/pull/45762#discussion_r875827186 would fix that though.

asbjornsabo commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1195,19 +1182,283 @@ static uint8_t discover_include_func(struct bt_conn *conn,
+	}
+#endif /* CONFIG_BT_MCC_OTS */
+
+#ifdef CONFIG_BT_MCC_OTS
+	/* Start discovery of included services to find OTS */
+	discover_included(conn);
+#else
+	/* If OTS is not configured, discovery ends here */
+	if (mcc_cb && mcc_cb->discover_mcs) {
+		mcc_cb->discover_mcs(conn, 0);
+	}
+#endif /* CONFIG_BT_MCC_OTS */
```
I believe all these services are tail-chained this way - at least vcs_client does more or less the same.
The discovery call triggers primary service discovery, which will trigger characteristic discovery (and subscription), which will trigger included service discovery, which will trigger discovery of characteristics of the included service.
And the callback is called when the chain ends.
The only difference here is that the subscriptions have been separated out as one additional step in the chain.

Thalley replied:
The difference is that this is not a callback, but rather a specific function that is called to do something. Semantically that is different to me, and there's no good argument not to rename this to better match what it does IMO.

asbjornsabo replied:
I see your point - that is certainly a difference.
To me, it is not that important.  The function was anyway called from (a) callback(s), and fitted in the chain.
But I have reworked the thing so that the function has a more descriptive name - `subscribe_next_mcs_char` and does only that.
The actual tail-chaining is moved to the subscription callback, to fit the existing pattern.  (The subscription callback has also been renamed to fit the other callbacks in the chain.)

Thalley commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1193,21 +1180,290 @@ static uint8_t discover_include_func(struct bt_conn *conn,
+			BT_DBG("Could not subscribe");
+			discovery_failure(conn, err);
+		}
+		return;
```
Hmm, could we use a generic `uint16_t handle` together with the `struct bt_gatt_subscribe_params *sub_params = NULL;` to move the 

```c
		err = do_subscribe(conn, cur_mcs_inst->player_name_handle, sub_params);
		if (err) {
			BT_DBG("Could not subscribe");
			discovery_failure(conn, err);
		}
```
part to a generic place (either with a goto or with a function) so that we only call `do_subscribe` one place?

Thalley replied:
Alternative move all the `if (handle && disc_param)` to a separate function that simply returns the `sub_params`?

Thalley commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1193,21 +1180,290 @@ static uint8_t discover_include_func(struct bt_conn *conn,
+		sub_params = &cur_mcs_inst->player_name_sub_params;
+		err = do_subscribe(conn, cur_mcs_inst->player_name_handle, sub_params);
+		if (err) {
+			BT_DBG("Could not subscribe");
```
```suggestion
			BT_DBG("Could not subscribe: %d", err);
```

Ditto for the other places

asbjornsabo replied:
Fixed everywhere, I believe.

asbjornsabo commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1049,6 +1032,14 @@ static uint8_t mcs_notify_handler(struct bt_conn *conn,
 	return BT_GATT_ITER_CONTINUE;
 }
 
+static void discovery_failure(struct bt_conn *conn, int err)
```
Could just check the error code, and reset the instance if not zero.

Thalley replied:
Do we not need to reset the MCS instance on discover complete? 

If we don't then we won't be able to call `bt_mcc_discover_mcs` again since it check for the pointer:

```c
int bt_mcc_discover_mcs(struct bt_conn *conn, bool subscribe)
{
	CHECKIF(!conn) {
		return -EINVAL;
	} else if (cur_mcs_inst) {
		return -EBUSY;
	}

```

asbjornsabo replied:
Changed.
Would like to postpone the reset functionality considerations, as it out of scope for this PR.

MariuszSkamra commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1225,8 +1493,8 @@ static uint8_t discover_mcs_char_func(struct bt_conn *conn,
-			sub_params = &cur_mcs_inst->player_name_sub_params;
-			sub_params->disc_params = &cur_mcs_inst->player_name_sub_disc_params;
+			/* Use discovery params pointer as subscription flag */
+			cur_mcs_inst->player_name_sub_params.disc_params = NULL;
```
Not related to this change, but you could save 32 bytes (16 characteristics * 2bytes per handle) if you could reuse `value_handle` stored in `bt_gatt_subscribe_params`

asbjornsabo replied:
I am not sure I follow.
The pointer to the discovery params can not be reused, as that is used as a persistent flag.
(This is anyway outside the scope of this PR, so I would like to take this as a separate issue.)

Thalley replied:
I think that his point is that `playback_speed_handle` is the same value as `playback_speed_sub_params.value_handle`, so we are basically duplicating that value for each characteristic we have a `struct bt_gatt_subscribe_params` for. 

I think this is outside the scope of this PR, and a general optimization we can have in LE Audio.

MariuszSkamra commented on subsys/bluetooth/audio/mcc.c:
```diff
@@ -1193,21 +1180,302 @@ static uint8_t discover_include_func(struct bt_conn *conn,
+{
+	int err;
+
+	discover_params.start_handle = cur_mcs_inst->start_handle;
```
Could it be an issue that`discover_params` are global (not per mcs instance?). What happens when connected to 2 devices in coordinated set?

asbjornsabo committed:
Bluetooth: Audio: Use common function for discovery complete

Use the same function at discovery complete, whether discovery failed
or succeeded.
Rename the discovery_failure() to discovery_complete(), and use that
also when discovery completes successfully.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented:
Rebased without changes on main, to get in fix for bug that broke the build.
